tonic/src/transport/endpoint.rs

```diff
@@ -252,3 +203,96 @@ impl fmt::Debug for Endpoint {
         f.debug_struct("Endpoint").finish()
     }
 }
+
+/// Configures TLS settings for endpoints.
+#[cfg(feature = "tls")]
+pub struct ClientTlsConfig {
+    provider: TlsProvider,
+    domain: Option<String>,
+    cert: Option<Certificate>,
+    #[cfg(feature = "openssl")]
+    openssl_raw: Option<openssl1::ssl::SslConnector>,
+    #[cfg(feature = "rustls")]
+    rustls_raw: Option<tokio_rustls::rustls::ClientConfig>,
+}
+
+#[cfg(feature = "tls")]
+impl fmt::Debug for ClientTlsConfig {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("ClientTlsConfig")
+            .field("provider", &self.provider)
+            .finish()
+    }
+}
+
+#[cfg(feature = "tls")]
+impl ClientTlsConfig {
+    /// Creates a new `ClientTlsConfig` backed by the specified provider. Enable the `openssl` or
+    /// `rustls` features of the `tonic` crate to use OpenSSL or Rustls respectively.
+    pub fn new(provider: TlsProvider) -> Self {
+        ClientTlsConfig {
+            provider,
+            domain: None,
+            cert: None,
+            #[cfg(feature = "openssl")]
+            openssl_raw: None,
+            #[cfg(feature = "rustls")]
+            rustls_raw: None,
+        }
+    }
+
+    /// Sets the domain name against which to verify the server's TLS certificate. If set to `None`
+    /// (the default), the address specified
+    pub fn domain_name(mut self, domain_name: impl Into<Option<String>>) -> Self {
+        self.domain = domain_name.into();
+        self
+    }
+
+    /// Sets the CA Certificate against which to verify the server's TLS certificate.
+    pub fn ca_certificate(mut self, ca_certificate: Certificate) -> Self {
+        self.cert = Some(ca_certificate);
+        self
+    }
+
+    /// Use options specified by the given `SslConnector` to configure TLS.
+    ///
+    /// This overrides all other TLS options set via other means.
+    #[cfg(feature = "openssl")]
+    pub fn openssl_connector(mut self, connector: openssl1::ssl::SslConnector) -> Self {
+        self.openssl_raw = Some(connector);
+        self
+    }
+
+    /// Use options specified by the given `ClientConfig` to configure TLS.
+    ///
+    /// This overrides all other TLS options set via other means.
+    #[cfg(feature = "rustls")]
+    pub fn rustls_client_config(mut self, config: tokio_rustls::rustls::ClientConfig) -> Self {
+        self.rustls_raw = Some(config);
+        self
+    }
+
+    fn tls_connector(self, uri: Uri) -> Option<TlsConnector> {
+        match self.provider {
+            TlsProvider::None => None,
```

Comment:
```suggestion
            TlsProvider::None => unreachable!("Created a tls provider with none enabled."),
```

Reply:
I've removed the `None` variant since it's no longer possible to construct it with the `with_openssl` and `with_rustls` functions being the entry point to the TLS builders.